### DISCUSSION TITLE
Use of Zim hierarchy with existing directories.

### DISCUSSION THREAD

**Discussion by tscahill** [zim-desktop-wiki/zim-desktop-wiki#1803] [Ideas]
I'm wondering if issue (Problems with "root document path" in properties #1146 ) and issue (Allow notebook document root to be relative to notebook folder location #198 ) shed light on something I'm trying to accomplish. I have a very large digital library, with a hierarchical directory structure built off that, with documents filed by topic folders. Because zim has such a light footprint, I would like to tie zim into this pre-existing directory, and use zim's capabilities to annotate and build context for the library contents. 

At this point, the library lives at /home/tim/Library. Here is a sample to a depth of 2 folders of the hierarchy the I've created.
├── 1 Software for reading ebooks
├── Health
│   ├── emergency 
│   ├── ill health
│   └── stress
├── How
│   ├── Ag-Garden
│   ├── Alt energy
│   ├── Boat
│   │ ....
│   │ ....
├── Humanities
│   ├── Anthropology
│   ├── Current Affairs
│   ├── Economics
│   │ ....
│   │ ....

The actual library hierachy is 44.9 GB, has 29,500 files, and 4,551 sub-folders. As the other issue tickets have noted, I'd rather have this library stay where it is, as it serves multiple purposes.

Zim's index structure is hung off an external (i.e, created and maintained by the OS' file system) directory structure. 
Is there a way that I could leverage this by having Zim use that directory structure as the structure of its index? I've tried pointing Zim at the folder, but of course, this causes the creation of a new subdir "home", independent of the existing hierarchy.

If I were to instead create the notebook in my root/home folder, and set 'home=Library' in notebook.zim's [Notebook] section, what could go wrong? The 'Add Notebook' dialog specifies that creating a new notebook requires specifying an empty folder, unless it is already a zim notebook folder. **_Why? Is this to do with initializing the index?_**

**Comment by introt** [comment]
While I lack enough familiarity with the codebase to give you the go-ahead, I can give you some practical advice as I've done a similar thing before.

1. You can create an empty notebook somewhere else, copy the `notebook.zim` file over & make the needed modifications to it by hand (ie. change `document_root` to `/home/tim/Library`; you might want to edit the `default_file_extension` as well)
2. To make Zim index your existing library hierarchy, you need two things: valid paths and dummy wiki pages
    * See eg. https://stackoverflow.com/questions/11809666/rename-files-using-regular-expression-in-linux for renaming your directories (spaces to underscores, get rid of `:`s, etc)
    * After completing the above, you can create the dummy pages very easily; something like `echo -e "Content-Type: text/x-zim-wiki\nWiki-Format: zim 0.6\nCreation-Date: 2021-12-13T06:00:00+02:00\n\nhello world" > template.txt && find -type d -exec cp template.txt {}.txt \; && rm template.txt` should do the trick (make sure you don't have a `template.txt` in the notebook root directory; see my reply below for an alternative without this caveat)
3. Open the notebook with Zim and hope for the best!

Remember to take and validate(!) backups before trying anything out. Good luck!

**Reply by introt** [reply]
I'm not sure whether the linked issues are relevant here; what I'm suggesting would result in your hierarchy ending up like this:

```
├── 1_Software_for_reading_ebooks/
├── 1_Software_for_reading_ebooks.txt
├── Health/
│   ├── emergency/
│   ├── emergency.txt
│   ├── ill_health/
│   ├── ill_health.txt
│   ├── stress/
│   └── stress.txt
├── Health.txt
...
```

**Reply by tscahill** [reply]
> * find -type d -exec -e "Content-Type: text/x-zim-wiki\nWiki-Format: zim 0.6\nCreation-Date: 2021-12-13T06:00:00+02:00\n\nhello world" {}.txt \;

Hey there. I've been puzzling over this compound command. I'm wondering if something is missing where "-e" is located. I can't find that parameter for -exec in the docs. I understand that the text within the double quotes are to become the contents of a new file, named by the contents of '{}' (the path to, and including, the current directory) with the file extention .txt.  

I'm trying to get this to work in a temporary folder structure
~/test
        /test
        
When I run the example command in ~/test, two errors are thrown.
`find: ‘-e’: No such file or directory 
find: ‘-e’: No such file or directory
`        
I expect the error is thrown twice because the command is executed once in each folder.

I've been pouring over the linux documentation to find a command that can take the quoted string and write it to the file that 'find' creates. 

Modifying the command as follows....
 `find -type d -exec "Content-Type: text/x-zim-wiki\nWiki-Format: zim 0.6\nCreation-Date:2021-12-13T06:00:00+02:00\n\nhello world" {}.txt \+`
 returns the error:
`find: In ‘-exec ... {} +’ the ‘{}’ must appear by itself, but you specified ‘{}.txt’`
When I remove the '.txt', 'find' thinks it is looking for the string in the double brackets.

**Reply by introt** [reply]
Hi @tscahill! Sorry, my bad - I tried to gather the necessary parts off the top of my head but didn't actually try the command line out. I've replaced it with a more readable one-liner that works; here's some terminal transcript:

```bash

$ tree
.
├── another
│   └── one
└── test
    └── me
        └── too

5 directories, 0 files
$ echo -e "Content-Type: text/x-zim-wiki\nWiki-Format: zim 0.6\nCreation-Date: 2021-12-13T06:00:00+02:00\n\nhello world" > template.txt && find -type d -exec cp template.txt {}.txt \; && rm template.txt
/dev/shm$ tree
.
├── another
│   ├── one
│   └── one.txt
├── another.txt
├── test
│   ├── me
│   │   ├── too
│   │   └── too.txt
│   └── me.txt
└── test.txt

5 directories, 5 files
$ cat test.txt 
Content-Type: text/x-zim-wiki
Wiki-Format: zim 0.6
Creation-Date: 2021-12-13T06:00:00+02:00

hello world
$ 
```

To make the original one-liner with no `cp`/temp file work, one would have to spawn a bunch of shells for the redirections; there are of course plenty of more obscure ways, but I see no point in getting into those here. Here's one working version of the original for educational purposes:

```bash
find -type d -execdir sh -c 'echo -e "Content-Type: text/x-zim-wiki\nWiki-Format: zim 0.6\nCreation-Date: 2021-12-13T06:00:00+02:00\n\nhello world" > '{}.txt \;
```

Notice the `-execdir` in place of the less secure `-exec`. It requires GNU(-compatible) `find` and can be used here interchangeably, but for our other variant we need an absolute path for the template; here, one is provided by `mktemp` via the `tmp` variable. It can be substituted by a path to an existing wiki file - just remember to remove the `echo`!

```bash
$ tmp=`mktemp`; echo -e "Content-Type: text/x-zim-wiki\nWiki-Format: zim 0.6\nCreation-Date: 2021-12-13T06:00:00+02:00\n\nhello world" > "$tmp" && find -type d -execdir cp "$tmp" {}.txt \; && rm "$tmp"
/dev/shm$ tree
.
├── another
│   ├── one
│   └── one.txt
├── another.txt
├── test
│   ├── me
│   │   ├── too
│   │   └── too.txt
│   └── me.txt
└── test.txt

5 directories, 5 files
$ cat test.txt 
Content-Type: text/x-zim-wiki
Wiki-Format: zim 0.6
Creation-Date: 2021-12-13T06:00:00+02:00

hello world
```

Best of luck with your endeavor!

**Comment by zimdoug** [comment]
As the original author of this issue:  
https://github.com/zim-desktop-wiki/zim-desktop-wiki  
(back even before it was on git)  

A primary motivation was exactly to make it where zim can integrate well with existing directory structures.
I haven't read the details of the discussion here, but I think that issue addresses the points well and seems has had some work.  
A couple of main problems at the moment are that zim uses a text file outside the directory being described, and that it won't navigate into subdirectories depths beyond ones described by those.    

That said, I think you can integrate zim the way it is at present, but you need to add a zim note matching the name of each directory. The Home directory is not a problem.  You can include other directories parallel to home.  So if you have Health, How and Home, in the Home note make type
```text
:Health
 
```
and enter, and you'll get a Health.txt in the top directory next to, not inside Home.  Click the link when it turns blue (after hitting enter)
Inside the Health note you'll want to type  
```
+emergency
```
and enter.  This creates a note as a sub-page of health, ie Health/emergency.txt and opens you up to now add notes in the Health/emergency subdirectory and so on.   For each note you can see attachments for that directory.  
But you must do the work to add note stubs for all directories or you can't see the ones below.
